File: src/main/scala/a_lists/P05.scala
package a_lists

object P05 {

  // P04. Reverse a list
  def reverse[A](xs: List[A]): List[A] = xs.reverse // built in

  def reverseWithFold[A](xs: List[A]): List[A] = xs.foldLeft(List[A]())((reversed, x) => x :: reversed)

  def reverseRecursive[A](xs: List[A], reversed: List[A] = Nil): List[A] = xs match {
    case Nil => reversed
    case h :: tail => reverseRecursive(tail, h :: reversed)
  }

}

File: ruby/a_lists/p05.rb
class P05

  # Reverse a list
  attr_reader :list

  def initialize(list)
    @list = list
  end

  def reverse
    list&.reverse
  end

  def reverse_alt
    return list if !list || list.length <= 1
    i = 0
    j = list.length - 1
    while i <= j
      list[i], list[j] = list[j], list[i]
      i += 1
      j -= 1
    end
    list
  end
end